### PR TITLE
[IMP] charts: improve design configuration

### PR DESCRIPTION
**Pull request by anhe-odoo**
## Description

This task aims to improve the design customization of the charts in o-spreadsheets, adding many customizable options like the title alignment/color/style, the axes configurations, the datasets colors, ...

## Related Task(s)

- Task [3792009](https://www.odoo.com/web#id=3792009&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## Review Checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: doc/tsdoc/README.md [hunk 1]
```diff
@@ -174,7 +174,7 @@ ___
 | Name | Type |
 | :------ | :------ |
 | `CellErrorLevel` | typeof `CellErrorLevel` |
-| `ChartColors` | typeof `ChartColors` |
+| `ColorGenerator` | typeof `ColorGenerator` |
```
Comment (anhe-odoo) [inline doc/tsdoc/README.md:177]: TODO : Need to be adapted in Odoo

File: src/components/side_panel/components/collapsible/side_panel_collapsible.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { Component } from "@odoo/owl";
+import { css } from "../../../helpers";
+
+const ANGLE_DOWN = /*xml*/ `
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 224 256">
+    <!--!Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2024 Fonticons, Inc.-->
```
Comment (anhe-odoo) [inline src/components/side_panel/components/collapsible/side_panel_collapsible.ts:6]: Remove licence once the svg will be optimized

File: tests/figures/chart/chart_plugin.test.ts [hunk 3]
```diff
@@ -1025,40 +1031,289 @@ describe("title", function () {
...
+    expect(getChartConfiguration("1").options?.plugins?.title?.display).toBe(true);
     updateChart(model, "1", { title: { content: "" } });
-    expect(
-      (model.getters.getChartRuntime("1") as BarChartRuntime).chartJsConfig.options?.plugins?.title
-        ?.display
-    ).toBe(false);
+    expect(getChartConfiguration("1").options?.plugins?.title?.display).toBe(false);
   });
+
+  test.each(["line", "bar", "pie", "combo", "waterfall", "scatter"] as const)(
```
Comment (anhe-odoo) [inline tests/figures/chart/chart_plugin.test.ts:1065]: Need to be in a previous commit

File: src/types/chart/chart.ts [hunk 4]
```diff
@@ -57,6 +57,37 @@ export interface DatasetValues {
   readonly data: any[];
 }
 
+export interface DatasetDesign {
+  readonly backgroundColor?: string;
+  readonly yAxisID?: string;
```
Comment (LucasLefevre) [inline src/types/chart/chart.ts:62]:
```suggestion
  readonly yAxisId?: string;
```

**Review by LucasLefevre — COMMENTED** [review]
First batch of comment :slightly_smiling_face:

File: src/types/chart/chart.ts (hunk 4 above)
Comment (anhe-odoo) [reply src/types/chart/chart.ts:62]: I used this case to make it the same as in ChartJS dataset config :/ But I can change if you really want

Comment (LucasLefevre) [reply src/types/chart/chart.ts:62]: Yes, I'd rather have consistent naming for our things

File: src/helpers/figures/charts/abstract_chart.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ import {
   ChartDefinition,
   ChartType,
   ExcelChartDefinition,
+  TitleDesign,
```
Comment (pro-odoo) [inline src/helpers/figures/charts/abstract_chart.ts:14]: `ChartTitle` ?

File: src/components/selection_input/selection_input.ts [hunk 6]
```diff
@@ -66,6 +66,7 @@ interface Props {
   class?: string;
   onSelectionChanged?: (ranges: string[]) => void;
   onSelectionConfirmed?: () => void;
+  initialColors?: Color[];
```
Comment (pro-odoo) [inline src/components/selection_input/selection_input.ts:69]: `colors` ?

**Review by pro-odoo — COMMENTED** [review]
some comments while scroll scroll

File: src/helpers/figures/charts/chart_common.ts [hunk 7]
```diff
@@ -388,3 +362,31 @@ export function getChartPositionAtCenterOfViewport(
...
+        weight: "bold" | "normal";
+      };
+      align: "start" | "center" | "end";
+    }
+  | undefined {
+  if (design?.title?.content) {
+    const { content, color, align, italic, bold } = design.title;
+    return {
+      display: true,
+      text: content,
```
Comment (LucasLefevre) [inline src/helpers/figures/charts/chart_common.ts:382]: To dynamically translate axis title extracted from the json files (in odoo)
```suggestion
      // axis title translatable terms are extracted from json files
      text: _t(content),
```

File: doc/tsdoc/README.md (hunk 1 above)
Comment (LucasLefevre) [inline doc/tsdoc/README.md:177]: revert this change in tsdoc. It's terrible outdated anyway

File: tests/figures/chart/gauge_rendering.test.ts [hunk 8]
```diff
@@ -53,11 +52,11 @@ describe("Gauge rendering config", () => {
 
   test("Chart title", () => {
     expect(getRenderingConfig(testRuntime).title).toEqual({
-      label: testRuntime.title,
+      label: testRuntime.title.content,
       fontSize: GAUGE_TITLE_FONT_SIZE,
       textPosition: {
         x: GAUGE_TITLE_PADDING_LEFT,
-        y: GAUGE_TITLE_PADDING_TOP + GAUGE_TITLE_FONT_SIZE,
+        y: NaN, //TODO : Is it possible to test this ?
```
Comment (LucasLefevre) [inline tests/figures/chart/gauge_rendering.test.ts:59]: NaN ?

File: tests/test_helpers/commands_helpers.ts [hunk 9]
```diff
@@ -161,15 +151,15 @@ export function createChart(
   return model.dispatch("CREATE_CHART", {
     id,
     sheetId,
+    //@ts-ignore WHY DO WE NEED THIS ???
```
Comment (LucasLefevre) [inline tests/test_helpers/commands_helpers.ts:154]: :eyes: we need to check this

File: src/migrations/data.ts [hunk 10]
```diff
@@ -388,6 +389,29 @@ const MIGRATIONS: Migration[] = [
       return data;
     },
   },
+  {
+    description: "transform chart data structure (2)",
```
Comment (LucasLefevre) [inline src/migrations/data.ts:393]: missing test for this in `data.test.ts`

File: src/migrations/data.ts [hunk 11]
```diff
@@ -388,6 +389,29 @@ const MIGRATIONS: Migration[] = [
...
+    from: 15,
+    to: 16,
+    applyMigration(data: any): any {
+      for (let sheet of data.sheets || []) {
+        for (let f in sheet.figures || []) {
```
Comment (LucasLefevre) [inline src/migrations/data.ts:398]:
```suggestion
      for (const sheet of data.sheets || []) {
        for (const f in sheet.figures || []) {
```

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+  DispatchResult,
+  SpreadsheetChildEnv,
+  TitleDesign,
+  UID,
+} from "../../../../../types";
+import { WaterfallChartDefinition } from "../../../../../types/chart/waterfall_chart";
+import { Section } from "../../../components/section/section";
+import { ChartTitle } from "../title/title";
+
+export interface axisDefinition {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:14]:
```suggestion
export interface AxisDefinition {
```

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.xml [hunk 13]
```diff
@@ -0,0 +1,30 @@
+<templates>
+  <t t-name="o-spreadsheet-GeneralDesignEditor">
+    <t t-set="chart_title">Chart Title</t>
+    <SidePanelCollapsible collapsedAtInit="false" name="'general'" class="general_collapsible">
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/general_design/general_design_editor.xml:4]: replace the manual name by a generated uuid in the component `SidePanelCollapsible` itself

**Review by LucasLefevre — COMMENTED** [review]
The components structure is much better now compared to the previous version with inheritance, nice :+1:

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    updateChart: Function,
+    slots: { type: Object, optional: true },
+  };
+  protected state!: GeneralDesignEditorState;
+
+  setup() {
+    const runtime = this.env.model.getters.getChartRuntime(this.props.figureId);
+    if (!runtime) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (anhe-odoo) [reply src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:50]: Nothing here, it was there in the previous version when everything was in the design side panel and I kept it here too, but it's not needed here

File: tests/figures/chart/gauge_rendering.test.ts (hunk 8 above)
Comment (anhe-odoo) [reply tests/figures/chart/gauge_rendering.test.ts:59]: Yes, it seems like we don't have any real size here in the test, so the computation of the position (base on the dimensions) give a nan

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+  axesList: AxisDefinition[];
+}
+
+export class AxisDesignEditor extends Component<Props, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-AxisDesignEditor";
+  static components = {
+    Section,
+    ChartTitle,
+  };
+  state = useState({ pickerOpened: false, index: 0, currentAxis: "x", activeTool: "" });
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:35]: seems unused
```suggestion
  state = useState({ currentAxis: "x" });
```

File: src/components/side_panel/chart/waterfall_chart/waterfall_chart_design_panel.ts [hunk 16]
```diff
@@ -1,22 +1,47 @@
...
+  };
+  static props = {
+    figureId: String,
+    definition: Object,
+    updateChart: Function,
+    canUpdateChart: { type: Function, optional: true },
+  };
 
-  state = useState({ pickerOpened: false });
+  state = useState({ pickerOpened: false, index: 0, currentAxis: "x", activeTool: "" });
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/waterfall_chart/waterfall_chart_design_panel.ts:43]: Looks like the state is only used in `closePicker` and `togglePicker` which are themselves unused
```suggestion
```

File: src/components/side_panel/chart/chart_with_axis/design_panel.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+export class ChartWithAxisDesignPanel extends Component<Props, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-ChartWithAxisDesignPanel";
+  static components = {
+    GeneralDesignEditor,
+    SidePanelCollapsible,
+    Section,
+    AxisDesignEditor,
+    RoundColorPicker,
+  };
+  state = useState({ pickerOpened: false, index: 0, currentAxis: "x", activeTool: "" });
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/chart_with_axis/design_panel.ts:34]: seems unused
```suggestion
  state = useState({ index: 0 });
```

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  get axisTitleStyle(): TitleDesign {
+    const axisDesign = this.props.definition.axesDesign?.[this.state.currentAxis] ?? {};
+    return {
+      color: "",
+      align: "center",
+      ...axisDesign.title,
+    };
+  }
+
+  updateAxisTitleColor(color: Color) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:46]: missing test

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+      ...axesDesign[this.state.currentAxis],
+      title: {
+        ...title,
+        bold: !title?.bold,
+      },
+    };
+    this.props.updateChart(this.props.figureId, { axesDesign });
+  }
+
+  toggleItalicAxisTitle() {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:72]: missing test

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+      ...axesDesign[this.state.currentAxis],
+      title: {
+        ...title,
+        italic: !title?.italic,
+      },
+    };
+    this.props.updateChart(this.props.figureId, { axesDesign });
+  }
+
+  updateAxisTitleAlignment(align: "left" | "center" | "right") {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:85]: missing test

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+    const axis = ev.target.value;
+    this.state.currentAxis = axis;
+  }
+
+  getAxisTitle() {
+    const axesDesign = this.props.definition.axesDesign ?? {};
+    return axesDesign[this.state.currentAxis]?.title.content || "";
+  }
+
+  updateAxisTitle(text: string) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:109]: missing test

File: src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+      title: {
+        ...title,
+        align,
+      },
+    };
+    this.props.updateChart(this.props.figureId, { axesDesign });
+    this.state.activeTool = "";
+  }
+
+  updateAxisEditor(ev) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:99]: missing test

File: src/components/side_panel/chart/chart_with_axis/design_panel.ts [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+  getDataSeries() {
+    const runtime = this.env.model.getters.getChartRuntime(this.props.figureId);
+    if (!runtime || !("chartJsConfig" in runtime)) {
+      return [];
+    }
+    return runtime.chartJsConfig.data.datasets.map((d) => d.label);
+  }
+
+  updateSerieEditor(ev) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/chart_with_axis/design_panel.ts:52]: missing test

File: src/components/side_panel/chart/chart_with_axis/design_panel.ts [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+    const chartId = this.props.figureId;
+    const selectedIndex = ev.target.selectedIndex;
+    const runtime = this.env.model.getters.getChartRuntime(chartId);
+    if (!runtime) {
+      return;
+    }
+    this.state.index = selectedIndex;
+  }
+
+  updateDataSeriesColor(color: string) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/chart_with_axis/design_panel.ts:62]: missing test

File: src/components/side_panel/chart/chart_with_axis/design_panel.ts [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  getDataSerieColor() {
+    const dataSets = this.props.definition.dataSets;
+    if (!dataSets?.[this.state.index]) return "";
+    const color = dataSets[this.state.index].backgroundColor;
+    return color ? toHex(color) : "";
+  }
+
+  updateDataSeriesAxis(ev) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/chart_with_axis/design_panel.ts:79]: missing test

File: src/components/side_panel/chart/chart_with_axis/design_panel.ts [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+    this.props.updateChart(this.props.figureId, { dataSets });
+  }
+
+  getDataSerieAxis() {
+    const dataSets = this.props.definition.dataSets;
+    if (!dataSets?.[this.state.index]) return "left";
+    return dataSets[this.state.index].yAxisId === "y1" ? "right" : "left";
+  }
+
+  updateDataSeriesLabel(ev) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/chart_with_axis/design_panel.ts:96]: missing test

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+
+  get titleStyle(): TitleDesign {
+    return {
+      color: "",
+      align: "left",
+      ...this.title,
+    };
+  }
+
+  updateChartTitleColor(color: Color) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:79]: missing test

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+    this.state = useState<GeneralDesignEditorState>({
+      activeTool: "",
+    });
+  }
+
+  get title(): TitleDesign {
+    return this.props.definition.title;
+  }
+
+  toggleDropdownTool(tool: string, ev: MouseEvent) {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:55]: missing test

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+    };
+  }
+
+  updateChartTitleColor(color: Color) {
+    const title = { ...this.title, color };
+    this.props.updateChart(this.props.figureId, { title } as T);
+    this.state.activeTool = "";
+  }
+
+  toggleBoldChartTitle() {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:85]: missing test

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+    let title = this.title;
+    title = { ...title, bold: !title.bold };
+    this.props.updateChart(this.props.figureId, { title } as T);
+  }
+
+  toggleItalicChartTitle() {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:91]: missing test

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+    let title = this.title;
+    title = { ...title, italic: !title.italic };
+    this.props.updateChart(this.props.figureId, { title } as T);
+  }
+
+  updateChartTitleAlignment(align: "left" | "center" | "right") {
```
Comment (LucasLefevre) [inline src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:97]: missing test

**Review by LucasLefevre — COMMENTED** [review]
`npm run test -- --coverage` shows a few interesting things :roll_eyes: :smile:

File: src/components/side_panel/chart/chart_with_axis/design_panel.ts (hunk 17 above)
Comment (anhe-odoo) [reply src/components/side_panel/chart/chart_with_axis/design_panel.ts:34]: State can be fully removed here :)

File: src/xlsx/extraction/chart_extractor.ts [hunk 32]
```diff
@@ -115,29 +106,55 @@ export class XlsxChartExtractor extends XlsxBaseExtractor {
...
+      })
+      .flat();
   }
 
   private extractScatterChartDatasets(chartElement: Element): ExcelChartDataset[] {
     return this.mapOnElements(
       { parent: chartElement, query: "c:ser" },
       (chartDataElement): ExcelChartDataset => {
+        let label = {};
+        const reference = this.extractChildTextContent(chartDataElement, "c:tx c:f");
```
Comment (LucasLefevre) [inline src/xlsx/extraction/chart_extractor.ts:147]: `c:xVal c:f`

File: src/xlsx/extraction/chart_extractor.ts [hunk 33]
```diff
@@ -115,29 +106,55 @@ export class XlsxChartExtractor extends XlsxBaseExtractor {
...
+        if (reference) {
+          label = { reference };
+        } else {
+          const text = this.extractChildTextContent(chartDataElement, "c:tx c:v");
```
Comment (LucasLefevre) [inline src/xlsx/extraction/chart_extractor.ts:151]: maybe check this also for scatter charts.

**Review by LucasLefevre — COMMENTED** [review]
we're getting close. Last review :)

File: src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+
+  updateTitle(newTitle: string) {
+    const title = { ...this.title, text: newTitle };
+    this.props.updateChart(this.props.figureId, { title } as T);
+  }
+
+  get titleStyle(): TitleDesign {
+    return {
+      color: "",
+      align: "left",
```
[the remark this answers is not in the document]
Comment (anhe-odoo) [reply src/components/side_panel/chart/building_blocks/general_design/general_design_editor.ts:74]: align is needed to provide the default alignment of the title, as it's different from the chart title (left) and the axis title (center). But color may be not needed here anymore :)

**Comment by rrahir** [comment]
@robodoo r+

yolo

**Comment by rrahir** [comment]
@robodoo rebase-ff
